### ISSUE TITLE
OnSMTPData event missing from DEBUG log

### ISSUE THREAD

**Issue by jimimaseye** [hmailserver/hmailserver#189]
(Tiny inconsistency. Its only minor but perhaps also a minor fix - just for completeness).

I note that when running un debug, the OnSMTPData 'event' doesnt get reported in the log (unlike the 'OnAcceptMessage' event, for example).

See example excerpt from a log here which shows the OnAcceptMessage event being to first and only event listed following the SMTP conversation:

> "SMTPD"   22496   39  "2016-07-08 08:46:19.451"   "127.0.0.1" "RECEIVED: DATA"
> "SMTPD"   22496   39  "2016-07-08 08:46:19.452"   "127.0.0.1" "SENT: 354 OK, send."
> "DEBUG"   16312   "2016-07-08 08:46:19.456"   "Adding task AsynchronousTask to work queue Asynchronous task queue"
> "DEBUG"   9260    "2016-07-08 08:46:19.457"   "Executing task AsynchronousTask in work queue Asynchronous task queue"
> "DEBUG"   9260    "2016-07-08 08:46:19.480"   "**Executing event OnAcceptMessage**"
> "DEBUG"   9260    "2016-07-08 08:46:19.538"   "Event completed"

**Comment by RvdHout** [comment]
Are you sure? I can see OnSMTPData in my logs

`"SMTPD"	2880	283	"2016-08-11 05:04:27.520"	"91.230.176.80"	"SENT: 220 Welcome to mail.domain.com ESMTP"
"SMTPD"	3232	283	"2016-08-11 05:04:27.536"	"91.230.176.80"	"RECEIVED: EHLO webgridf080.emsecure.net"
"DEBUG"	3232	"2016-08-11 05:04:27.536"	"Executing event OnHELO"
"DEBUG"	3232	"2016-08-11 05:04:27.536"	"Event completed"
"SMTPD"	3232	283	"2016-08-11 05:04:27.536"	"91.230.176.80"	"SENT: 250-ip3e832edd.speed.planet.nl[nl]250-SIZE[nl]250-STARTTLS[nl]250 HELP"
"SMTPD"	964	283	"2016-08-11 05:04:27.551"	"91.230.176.80"	"RECEIVED: MAIL FROM:<info@kruidvat.emsecure.net>"
"DEBUG"	964	"2016-08-11 05:04:27.797"	"Spam test: SpamTestHeloHost, Score: 0"
"DEBUG"	964	"2016-08-11 05:04:27.813"	"Spam test: SpamTestMXRecords, Score: 0"
"DEBUG"	964	"2016-08-11 05:04:27.844"	"Spam test: SpamTestSPF, Score: 0"
"DEBUG"	964	"2016-08-11 05:04:27.844"	"Total spam score: 0"
"SMTPD"	964	283	"2016-08-11 05:04:27.860"	"91.230.176.80"	"SENT: 250 OK"
"SMTPD"	3232	283	"2016-08-11 05:04:27.876"	"91.230.176.80"	"RECEIVED: RCPT TO:<user@domain.com>"
"DEBUG"	3232	"2016-08-11 05:04:27.876"	"SPF passed, skipping greylisting."
"SMTPD"	3232	283	"2016-08-11 05:04:27.876"	"91.230.176.80"	"SENT: 250 OK"
"SMTPD"	2880	283	"2016-08-11 05:04:27.891"	"91.230.176.80"	"RECEIVED: DATA"
"DEBUG"	2880	"2016-08-11 05:04:27.891"	"Executing event OnSMTPData"
"DEBUG"	2880	"2016-08-11 05:04:27.891"	"Event completed"
"SMTPD"	2880	283	"2016-08-11 05:04:27.891"	"91.230.176.80"	"SENT: 354 OK, send."

`

**Comment by OutbackMatt** [comment]
I don't see OnSMTPData being logged either.
I know it runs because I have a custom log that runs from that.

I also don't have an OnHELO event  :)

**Comment by RvdHout** [comment]
I use a custom build based on 5.6.5-B2367 with support for the OnHELO event trigger, eg: 
https://www.hmailserver.com/forum/viewtopic.php?p=188753#p188753

In ScriptServer.cpp I edited in the changes for 5.7 so maybe that is why the OnSMTPData trigger is displayed on my installation

 **ScriptServer.cpp** 

`// Copyright (c) 2010 Martin Knafve / hMailServer.com.  
// http://www.hmailserver.com

#include "StdAfx.h"


#include "../../hMailServer/hMailServer.h"


#include "ScriptServer.h"
#include "ScriptSite.h"
#include "ScriptObjectContainer.h"


#ifdef _DEBUG
#define DEBUG_NEW new(_NORMAL_BLOCK, __FILE__, __LINE__)
#define new DEBUG_NEW
#endif

namespace HM
{
	ScriptServer::ScriptServer(void) :
		has_on_client_connect_(false),
		has_on_accept_message_(false),
		has_on_deliver_message_(false),
		has_on_backup_completed_(false),
		has_on_backup_failed_(false),
		has_on_delivery_start_(false),
		has_on_error_(false),
		has_on_delivery_failed_(false),
		has_on_external_account_download_(false),
		has_on_smtpdata_(false),
		has_on_helo_(false)
	{

	}

	ScriptServer::~ScriptServer(void)
	{

	}

	String
		ScriptServer::GetCurrentScriptFile() const
	{
		String sEventsDir = IniFileSettings::Instance()->GetEventDirectory();

		String sScriptLanguage = Configuration::Instance()->GetScriptLanguage();

		String sExtension;
		if (sScriptLanguage == _T("VBScript"))
			sExtension = "vbs";
		else if (sScriptLanguage == _T("JScript"))
			sExtension = "js";
		else
			sExtension = "";

		String sFileName = sEventsDir + "\\EventHandlers." + sExtension;

		return sFileName;
	}

	void
		ScriptServer::LoadScripts()
	{
		try
		{
			script_language_ = Configuration::Instance()->GetScriptLanguage();

			if (script_language_ == _T("VBScript"))
				script_extension_ = "vbs";
			else if (script_language_ == _T("JScript"))
				script_extension_ = "js";
			else
				script_extension_ = "";

			String sCurrentScriptFile = GetCurrentScriptFile();

			// Load the script file from disk
			script_contents_ = FileUtilities::ReadCompleteTextFile(sCurrentScriptFile);

			// Do a syntax check before loading it.
			String sMessage = ScriptServer::CheckSyntax();
			if (!sMessage.IsEmpty())
			{
				ErrorManager::Instance()->ReportError(ErrorManager::High, 5016, "ScriptServer::LoadScripts", sMessage);
				return;
			}

			// Determine which functions are available.
			has_on_client_connect_ = DoesFunctionExist_("OnClientConnect");
			has_on_accept_message_ = DoesFunctionExist_("OnAcceptMessage");
			has_on_deliver_message_ = DoesFunctionExist_("OnDeliverMessage");
			has_on_backup_completed_ = DoesFunctionExist_("OnBackupCompleted");
			has_on_backup_failed_ = DoesFunctionExist_("OnBackupFailed");
			has_on_delivery_start_ = DoesFunctionExist_("OnDeliveryStart");
			has_on_error_ = DoesFunctionExist_("OnError");
			has_on_delivery_failed_ = DoesFunctionExist_("OnDeliveryFailed");
			has_on_external_account_download_ = DoesFunctionExist_("OnExternalAccountDownload");
			has_on_smtpdata_ = DoesFunctionExist_("OnSMTPData");
			has_on_helo_ = DoesFunctionExist_("OnHELO");

		}
		catch (...)
		{
			ErrorManager::Instance()->ReportError(ErrorManager::High, 5017, "ScriptServer::LoadScripts", "An exception was thrown when loading scripts.");
		}
	}


	String
		ScriptServer::CheckSyntax()
	{
		String sEventsDir = IniFileSettings::Instance()->GetEventDirectory();

		String sScriptLanguage = Configuration::Instance()->GetScriptLanguage();
		String sScriptExtension;
		if (sScriptLanguage == _T("VBScript"))
			sScriptExtension = "vbs";
		else if (sScriptLanguage == _T("JScript"))
			sScriptExtension = "js";
		else
			sScriptExtension = "";

		String sErrorMessage;

		// Compile the scripts.
		sErrorMessage = Compile_(sScriptLanguage, sEventsDir + "\\EventHandlers." + sScriptExtension);
		if (!sErrorMessage.IsEmpty())
			return sErrorMessage;

		return String("");
	}

	bool
		ScriptServer::DoesFunctionExist_(const String &sProcedure)
	{
		// Create an instance of the script engine and execute the script.
		CComObject<CScriptSiteBasic>* pBasic;
		CComObject<CScriptSiteBasic>::CreateInstance(&pBasic);

		CComQIPtr<IActiveScriptSite> spUnk;

		if (!pBasic)
		{
			ErrorManager::Instance()->ReportError(ErrorManager::High, 5017, "ScriptServer::FireEvent", "Failed to create instance of script site.");
			return false;
		}

		spUnk = pBasic;
		pBasic->Initiate(script_language_, NULL);
		pBasic->AddScript(script_contents_);
		pBasic->Run();
		bool bExists = pBasic->ProcedureExists(sProcedure);
		pBasic->Terminate();

		return bExists;
	}


	String
		ScriptServer::Compile_(const String &sLanguage, const String &sFilename)
	{
		String sContents = FileUtilities::ReadCompleteTextFile(sFilename);

		if (sContents.IsEmpty())
			return "";

		// Create an instance of the script engine and execute the script.
		CComObject<CScriptSiteBasic>* pBasic;
		CComObject<CScriptSiteBasic>::CreateInstance(&pBasic);

		CComQIPtr<IActiveScriptSite> spUnk;

		if (!pBasic)
			return "ScriptServer:: Failed to create instance of script site.";

		spUnk = pBasic; // let CComQIPtr tidy up for us
		pBasic->Initiate(sLanguage, NULL);
		// pBasic->SetObjectContainer(pObjects);
		pBasic->AddScript(sContents);
		pBasic->Run();
		pBasic->Terminate();

		String sErrorMessage = pBasic->GetLastError();

		if (!sErrorMessage.IsEmpty())
			sErrorMessage = "File: " + sFilename + "\r\n" + sErrorMessage;

		return sErrorMessage;

	}

	void
		ScriptServer::FireEvent(Event e, const String &sEventCaller, std::shared_ptr<ScriptObjectContainer> pObjects)
	{
		if (!Configuration::Instance()->GetUseScriptServer())
			return;

		// JDR: stores the name of the method that is fired in the script. http://www.hmailserver.com/forum/viewtopic.php?f=2&t=25497
		String event_name = _T("Unknown");

		switch (e)
		{
		case EventOnClientConnect:
			if (!has_on_client_connect_)
				return;
			event_name = _T("OnClientConnect");
			break;
		case EventOnAcceptMessage:
			if (!has_on_accept_message_)
				return;
			event_name = _T("OnAcceptMessage");
			break;
		case EventOnDeliverMessage:
			if (!has_on_deliver_message_)
				return;
			event_name = _T("OnDeliverMessage");
			break;
		case EventOnBackupCompleted:
			if (!has_on_backup_completed_)
				return;
			event_name = _T("OnBackupCompleted");
			break;
		case EventOnBackupFailed:
			if (!has_on_backup_failed_)
				return;
			event_name = _T("OnBackupFailed");
			break;
		case EventOnError:
			if (!has_on_error_)
				return;
			event_name = _T("OnError");
			break;
		case EventOnDeliveryStart:
			if (!has_on_delivery_start_)
				return;
			event_name = _T("OnDeliveryStart");
			break;
		case EventOnDeliveryFailed:
			if (!has_on_delivery_failed_)
				return;
			event_name = _T("OnDeliveryFailed");
			break;
		case EventOnExternalAccountDownload:
			if (!has_on_external_account_download_)
				return;
			event_name = _T("OnExternalAccountDownload");
			break;
		case EventOnSMTPData:
			if (!has_on_smtpdata_)
				return;
			event_name = _T("OnSMTPData");
			break;
		case EventOnHELO:
			if (!has_on_helo_)
				return;
			event_name = _T("OnHELO");
			break;
		case EventCustom:
			break;
		default:
		{
			return;
		}
		}

		LOG_DEBUG("Executing event " + event_name);

		String sScript;

		// Build the script.
		if (script_language_ == _T("VBScript"))
			sScript = script_contents_ + "\r\n\r\n" + "Call " + sEventCaller + "\r\n";
		else if (script_language_ == _T("JScript"))
			sScript = script_contents_ + "\r\n\r\n" + sEventCaller + ";\r\n";

		CComObject<CScriptSiteBasic>* pBasic;
		CComObject<CScriptSiteBasic>::CreateInstance(&pBasic);
		CComQIPtr<IActiveScriptSite> spUnk;

		if (!pBasic)
		{
			ErrorManager::Instance()->ReportError(ErrorManager::High, 5018, "ScriptServer::FireEvent", "Failed to create instance of script site.");
			return;
		}

		spUnk = pBasic; // let CComQIPtr tidy up for us
		pBasic->Initiate(script_language_, NULL);
		pBasic->SetObjectContainer(pObjects);
		pBasic->AddScript(sScript);
		pBasic->Run();
		pBasic->Terminate();

		LOG_DEBUG("Event completed");
	}
}`

**Comment by martinknafve** [comment]
Realize that this is an old issue, but I tested this now and I can see *Executing event OnSMTPData* in the hMailServer log file. I did fix some issues in this code around the time of the above discussion, so I'm guessing the issue is resolved. I checked the code as well, and I didn't see any obvious reason that OnSMTPData wouldn't be logged. The code which logs "Executing event X" is generic and not specific to any event.
